File: scratch/test_fuzzy_dedupe.js
const crypto = require('crypto');

// Python normalize_title equivalent in JS for test comparison
function normalizeTitle(title, articleUrl) {
  let t = title.toLowerCase();
  
  // Normalize accents and diacritics
  t = t.normalize("NFD").replace(/[\u0300-\u036f]/g, "");
  
  // Keep only alphanumeric and spaces
  t = t.replace(/[^a-z0-9\s]/g, " ");
  
  // Remove common Spanish stopwords
  const stopwords = new Set([
    "el", "la", "los", "las", "un", "una", "unos", "unas",
    "de", "del", "en", "para", "por", "con", "sin", "sobre",
    "y", "o", "u", "a", "al", "que", "se", "su", "sus", "como"
  ]);
  
  const words = t.split(/\s+/).filter(w => w && !stopwords.has(w));
  let slug = words.join("-");
  
  if (!slug) {
    slug = crypto.createHash('md5').update(articleUrl).digest('hex').slice(0, 12);
  }
  
  return slug.slice(0, 100);
}

// Levenshtein functions from workers/src/utils.js
function levenshteinDistance(a, b) {
  if (a === b) return 0;
  if (a.length === 0) return b.length;
  if (b.length === 0) return a.length;

  let prev = Array.from({ length: b.length + 1 }, (_, i) => i);

  for (let i = 1; i <= a.length; i++) {
    const curr = [i];
    for (let j = 1; j <= b.length; j++) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      curr[j] = Math.min(
        curr[j - 1] + 1,        // insertion
        prev[j]     + 1,        // deletion
        prev[j - 1] + cost,     // substitution
      );
    }
    prev = curr;
  }

  return prev[b.length];
}

function levenshteinRatio(a, b) {
  if (a === b) return 0;
  const maxLen = Math.max(a.length, b.length);
  if (maxLen === 0) return 0;
  return Number((levenshteinDistance(a, b) / maxLen).toFixed(4));
}

// Test cases
const titles = [
  ["Hacienda lanzará una inspección masiva de autónomos en junio", "https://url1.com"],
  ["Hacienda lanza inspección masiva a los autónomos en junio", "https://url2.com"],
  ["La Agencia Tributaria inspeccionará masivamente a autónomos el mes que viene", "https://url3.com"],
  ["Hacienda aprueba ayudas de 200 euros para pymes", "https://url4.com"],
  ["Ayudas de 200 euros de Hacienda para pymes y autónomos", "https://url5.com"],
];

console.log("--- Slugs generated ---");
const slugs = [];
for (const [t, url] of titles) {
  const slug = normalizeTitle(t, url);
  slugs.push(slug);
  console.log(`Title: ${t}\nSlug : ${slug}\n`);
}

console.log("--- Similarity checking ---");
for (let i = 0; i < slugs.length; i++) {
  for (let j = i + 1; j < slugs.length; j++) {
    const ratio = levenshteinRatio(slugs[i], slugs[j]);
    const isDup = ratio <= 0.18;
    console.log(`Slug A: ${slugs[i]}`);
    console.log(`Slug B: ${slugs[j]}`);
    console.log(`Levenshtein ratio: ${ratio.toFixed(4)} (Similarity: ${((1 - ratio) * 100).toFixed(1)}%) | Duplicate: ${isDup}`);
    console.log("-".repeat(40));
  }
}

File: scratch/test_fuzzy_dedupe.py
import hashlib
import unicodedata

def normalize_title(title: str, article_url: str) -> str:
    """
    Normalizes a title to generate a stable, readable slug for duplicate checking.
    """
    # 1. Lowercase
    t = title.lower()
    # 2. Remove accents and diacritics
    t = "".join(
        c for c in unicodedata.normalize("NFD", t)
        if unicodedata.category(c) != "Mn"
    )
    # 3. Replace non-alphanumeric characters with spaces
    t = "".join(c if c.isalnum() else " " for c in t)
    # 4. Remove common Spanish stopwords
    stopwords = {
        "el", "la", "los", "las", "un", "una", "unos", "unas",
        "de", "del", "en", "para", "por", "con", "sin", "sobre",
        "y", "o", "u", "a", "al", "que", "se", "su", "sus", "como"
    }
    words = [w for w in t.split() if w not in stopwords]
    # 5. Join words with hyphens
    slug = "-".join(words)
    # 6. Fallback to URL hash if slug is empty
    if not slug:
        slug = hashlib.md5(article_url.encode()).hexdigest()[:12]
    return slug[:100]

def levenshtein_distance(a: str, b: str) -> int:
    if a == b: return 0
    if len(a) == 0: return len(b)
    if len(b) == 0: return len(a)
    
    prev = list(range(len(b) + 1))
    for i in range(1, len(a) + 1):
        curr = [i]
        for j in range(1, len(b) + 1):
            cost = 0 if a[i - 1] == b[j - 1] else 1
            curr.append(min(
                curr[j - 1] + 1,      # insertion
                prev[j] + 1,          # deletion
                prev[j - 1] + cost    # substitution
            ))
        prev = curr
    return prev[len(b)]

def levenshtein_ratio(a: str, b: str) -> float:
    if a == b: return 0.0
    max_len = max(len(a), len(b))
    if max_len == 0: return 0.0
    return round(levenshtein_distance(a, b) / max_len, 4)

# Test cases
titles = [
    ("Hacienda lanzará una inspección masiva de autónomos en junio", "https://url1.com"),
    ("Hacienda lanza inspección masiva a los autónomos en junio", "https://url2.com"),
    ("La Agencia Tributaria inspeccionará masivamente a autónomos el mes que viene", "https://url3.com"),
    ("Hacienda aprueba ayudas de 200 euros para pymes", "https://url4.com"),
    ("Ayudas de 200 euros de Hacienda para pymes y autónomos", "https://url5.com"),
]

print("--- Slugs generated ---")
slugs = []
for t, url in titles:
    slug = normalize_title(t, url)
    slugs.append(slug)
    print(f"Title: {t}\nSlug : {slug}\n")

print("--- Similarity checking ---")
for i in range(len(slugs)):
    for j in range(i + 1, len(slugs)):
        ratio = levenshtein_ratio(slugs[i], slugs[j])
        is_dup = ratio <= 0.18
        print(f"Slug A: {slugs[i]}")
        print(f"Slug B: {slugs[j]}")
        print(f"Levenshtein ratio: {ratio:.4f} (Similarity: {(1 - ratio) * 100:.1f}%) | Duplicate: {is_dup}")
        print("-" * 40)
